### PR TITLE
GUACAMOLE-1204: Migrate Guacamole.Mouse from legacy event handlers to the new event stack.

### PR DESCRIPTION
**Pull request by mike-jumper**
This change migrates `Guacamole.Mouse` from the old-style `onthing` event handlers:

```javascript
mouse.onmousedown =
mouse.onmouseup   =
mouse.onmousemove = function sendMouseState(state) {
    guac.sendMouseState(state);
};
```

to the new event stack:

```javascript
mouse.onEach(['mousedown', 'mouseup', 'mousemove'], function sendMouseEvent(e) {                
    guac.sendMouseState(e.state);
});
```

To keep things reasonable for review, I've limited this body of changes to only those necessary to migrate `Guacamole.Mouse`. Other PRs will follow for other eventy parts of the Guacamole JavaScript API. ;)

### CODE REVIEW

File: guacamole-common-js/src/main/webapp/modules/Mouse.js [hunk 1]
```diff
@@ -503,16 +425,355 @@ Guacamole.Mouse.State = function State(template) {
...
+     * {@link Guacamole.Mouse.Touchscreen}), or may be programmatically
+     * generated through [dispatch()]{@link Guacamole.Mouse.Event.Target#dispatch},
+     * [press()]{@link Guacamole.Mouse.Event.Target#press}, or
+     * [click()]{@link Guacamole.Mouse.Event.Target#click}.
+     *
+     * @event Guacamole.Mouse.Event.Target#mousedown
+     * @param {Guacamole.Mouse.Event} event
+     *     The mousedown event that was fired.
+     */
+
```
Comment (necouchman) [inline guacamole-common-js/src/main/webapp/modules/Mouse.js:578]: I'm assuming that these are just for documentation purposes, with the `@event` tag, but just wanted to make sure that it's expected that there isn't any actual code being documented, here?

File: guacamole-common-js/src/main/webapp/modules/Mouse.js [hunk 2]
```diff
@@ -503,16 +425,355 @@ Guacamole.Mouse.State = function State(template) {
...
+     * Clicks (presses and releases) the given mouse button. Valid button
+     * names are defined by {@link Guacamole.Mouse.State.Buttons} and
+     * correspond to the button-related properties of
+     * {@link Guacamole.Mouse.State}.
+     *
+     * @fires Guacamole.Mouse.Event.Target#mousedown
+     * @fires Guacamole.Mouse.Event.Target#mouseup
+     *
+     * @param {String} button
+     *     The name of the mouse button to release, as defined by
```
Comment (necouchman) [inline guacamole-common-js/src/main/webapp/modules/Mouse.js:675]: The name of the mouse button to _click_, I think?

File: guacamole-common-js/src/main/webapp/modules/Mouse.js [hunk 3]
```diff
@@ -503,16 +425,355 @@ Guacamole.Mouse.State = function State(template) {
...
+    /**
+     * Presses the given mouse button, if it isn't already pressed. Valid
+     * button names are defined by {@link Guacamole.Mouse.State.Buttons} and
+     * correspond to the button-related properties of
+     * {@link Guacamole.Mouse.State}.
+     *
+     * @fires Guacamole.Mouse.Event.Target#mousedown
+     *
+     * @param {String} button
+     *     The name of the mouse button to release, as defined by
```
Comment (necouchman) [inline guacamole-common-js/src/main/webapp/modules/Mouse.js:630]: The name of the mouse button to _press_, perhaps?

**Review by necouchman — CHANGES_REQUESTED** [review]
Minor copy-pasta issues, but other than that looks okay.

I assume these changes are all backward compatible to the extent that they won't break any specific browsers (thinking of older IE versions)?

**Comment by mike-jumper** [comment]
> I assume these changes are all backward compatible to the extent that they won't break any specific browsers (thinking of older IE versions)?

Yep. In addition to the usual modern browsers, I've specifically tested these changes against IE10 and IE11.

This also shouldn't break older usages of guacamole-common-js thanks to the `invokeLegacyHandler()` logic provided by `Guacamole.Event`.

File: guacamole-common-js/src/main/webapp/modules/Mouse.js (hunk 1 above)
Comment (mike-jumper) [reply guacamole-common-js/src/main/webapp/modules/Mouse.js:578]: Yep, that's correct. These will be picked up by JSDoc and documented as events.

**Comment by mike-jumper** [comment]
OK - rebased to resolve the pasta.

**Pull request merged**